master/internal/rm/multirm/multirm.go

```diff
@@ -0,0 +1,435 @@
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/determined-ai/determined/master/internal/config"
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/rm"
+	"github.com/determined-ai/determined/master/internal/rm/agentrm"
+	"github.com/determined-ai/determined/master/internal/rm/kubernetesrm"
+	"github.com/determined-ai/determined/master/internal/sproto"
+	"github.com/determined-ai/determined/master/pkg/aproto"
+	"github.com/determined-ai/determined/master/pkg/command"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/agentv1"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+	"github.com/determined-ai/determined/proto/pkg/jobv1"
+	"github.com/determined-ai/determined/proto/pkg/resourcepoolv1"
+)
+
+// ErrRPNotDefined returns a detailed error if a resource manager/pool combination isn't found in the MultiRMRouter map.
+func ErrRPNotDefined(rm, rp string) error {
+	return fmt.Errorf("resource pool %s not found in resource manager %s", rp, rm)
+}
+
+// MultiRMRouter tracks all resource managers in the system.
+type MultiRMRouter struct {
+	mu     sync.Mutex
+	rms    map[string]rm.ResourceManager
+	syslog *logrus.Entry
+}
+
+// New creates a new Multi-RM router with the master set-up.
+func New(
+	db *db.PgDB,
+	echo *echo.Echo,
+	rmConfigs []*config.ResourceManagerWithPoolsConfig,
+	tcd *model.TaskContainerDefaultsConfig,
+	opts *aproto.MasterSetAgentOptions,
+	cert *tls.Certificate,
+) rm.ResourceManager {
+	if len(rmConfigs) <= 1 {
+		config := rmConfigs[0]
+		switch {
+		case config.ResourceManager.AgentRM != nil:
+			return agentrm.New(db, echo, config, opts, cert)
+		case config.ResourceManager.KubernetesRM != nil:
+			return kubernetesrm.New(db, config, tcd, opts, cert)
+		default:
+			panic("no expected resource manager config is defined")
+		}
+	}
+
+	rms := map[string]rm.ResourceManager{}
+	for _, cfg := range rmConfigs {
+		c := cfg
+		switch {
+		case c.ResourceManager.AgentRM != nil:
+			// TODO (multirm): what to do if Name is ""?
+			rms[c.ResourceManager.AgentRM.Name] = agentrm.New(db, echo, c, opts, cert)
+		case c.ResourceManager.KubernetesRM != nil:
+			// TODO (multirm): ditto.
+			rms[c.ResourceManager.KubernetesRM.Name] = kubernetesrm.New(db, c, tcd, opts, cert)
+		default:
+			panic("no expected resource manager config is defined")
+		}
+	}
+	return &MultiRMRouter{
+		rms:    rms,
+		syslog: logrus.WithField("component", "resource-router"),
+	}
+}
+
+// GetAllocationSummaries returns the allocation summaries for all resource pools across all resource managers.
+func (m *MultiRMRouter) GetAllocationSummaries() (
+	map[model.AllocationID]sproto.AllocationSummary,
+	error,
+) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var eg errgroup.Group
+	var mu sync.Mutex
+	res := map[model.AllocationID]sproto.AllocationSummary{}
+
+	for _, rm := range m.rms {
+		r := rm
```

Comment:
What is the purpose of this line? Does it keep rm in scope? (I'm not familiar with this)

Reply:
tldr; "loop variable captured by func literal", so it's good practice to shadow the variable before passing it in.

references: https://go.dev/wiki/CommonMistakes, https://fuchsoria.medium.com/shadowing-or-how-to-fix-loop-variable-i-captured-by-func-literal-f365c0ee984e